onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,171 @@
+       |   |
+       |   '-----------------------------------> dY/dH (1st output of Gradient)
+       |
+       '---------------------------------------> dY/dZ (2nd output of Gradient)
+```
+
+It is possible to represent high-order differentiation using Gradient operator.
+An example for linear model is
+
+```
+W --> Gemm --> Y --> Loss --> O
+       ^              ^
+       |              |
+       X              L
+```
+
+To compute the 2nd order derivative of O with respect to W (denoted by
+d^2O/dW^2), one can do
+
+```
+W --> Gemm --> Y --> Loss --> O
+|      ^              ^
+|      |              |
+|      X              L
+|      |
+|      |
++------+--> Gradient(xs=["X", "W"], y=["Y"]) ---> dO/dX (1st output of Gradient)
+|      |      |
+|      |      '---> dO/dW (2nd output of Gradient)
+|      v
+'---> Gradient(xs=["X", "W"], y=["dO/dW"]) ---> d(dO/dW)dX (1st output of
+       |                                          Gradient)
+       |
+       |
+       '---> d^2O/dW^2 (2nd output of Gradient)
+```
+
+As mentioned above, the attributes "xs" and "y" are used to identify a graph,
+and we can feed different tensors to the identified graph. For,
+example, one can compute the gradient of Y with respect to H by substituting
+Y_1 into Y and H_1 into H.
+
+```
+W --> Conv --> H --> Gemm --> Y
+       ^              ^
+       |              |
+       X              Z
+
+           Z_1 (the 2nd input of Gradient)
+           |
+           v
+W_1 --> Gradient(xs=["H", "Z"], y=["Y"]) ---> dY/dX when Y = Y_1
+         |   |
+         |   '-----------------------------------> dY/dW (2nd output of Gradient)
+         |
+         '---------------------------------------> dY/dZ (3rd output of Gradient)
+```
+
+This graph also implies that one can identify a gradient graph once and reuse
+it multiple times with different input values.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Gradient,
+    12,
+    OpSchema()
+        .SetDoc(Gradient_ver12_doc)
+        .Input(
+            0,
+            "Input",
+            "The values fed into graph identified by the attributes. "
+            "The i-th input is the value of the i-th tensor specified in the "
+            "attribute \"xs\".",
+            "V",
+            OpSchema::Variadic,
+            false)
+        .Output(
+            0,
+            "Output",
+            "The gradient of the attribute \"y\" with respect to \"xs\". "
+            "The i-th output is the gradient of \"y\" with respect to the "
+            "i-th tensor specified in the attribute \"xs\".",
+            "V",
+            OpSchema::Variadic,
+            false)
+        .Attr(
+            "xs",
+            "Input tensor names of the differentiated sub-graph. It "
+            "contains and only contains the necessary inputs of a "
+            "(sub-)graph. Variables (usually called intermediate "
+            "variables) which can be generated by inputs cannot be "
+            "included in this attribute.",
```

Comment:
Is it correct to say the following? If xs=[X_1, ..., X_m], then no X_i can depend on X_j (i <> j): that is, there cannot be a dependence-path in the graph from X_j to X_i? Furthermore, do we need to say the list in xs must be complete? If we have Y = X_1 + X_2, is it valid or invalid to specify xs=["X_1"]?

Reply:
Yes. The inputs named in `xs` must be independent variables; that is, they cannot be upstream variable of each other.

The list must be complete. Otherwise, we cannot finish a forward path due to missing input (in your example, we cannot compute `Y` without `X_2`).